File: .config/nvim/testing/sample.js
// Sample JavaScript file with various issues for testing

// Variables with potential issues
let name = "World";
const count = 5;
var unusedVar = "This variable is not used";
let anotherUnused = 42;
var globalVar = "Should use const/let";

// Function with potential issues
function greet(name) {
    console.log(`Hello, ${name}!`);
    console.log(`Count is: ${count}`);

    // Use of undeclared variable
    console.log(undeclaredVar);

    // Missing semicolon
    let message = "This message is missing a semicolon"

    // Unused parameter
    const unusedParam = (param1, param2, param3) => {
        return param1;
    };

    return message;
}

// Arrow function with potential issues
const processData = (data) => {
    return data.map(item => {
        return {
            id: item.id,
            name: item.name.toUpperCase(),
            processed: true
        };
    });
};

// Object with potential issues
const config = {
    apiUrl: 'https://api.example.com',
    timeout: 5000,
    retries: 3,
    // Trailing comma
};

// Object with duplicate keys
const badConfig = {
    url: 'https://example.com',
    timeout: 3000,
    url: 'https://backup.com', // Duplicate key
};

// Array with potential issues
const items = [1, 2, 3, 4, 5];
const mixedArray = ['string', 42, true, null, undefined, { key: 'value' }];
const emptyArray = [];

// Function that's never called
const neverCalled = () => {
    console.log('This will never be called');
};

// Conditional with potential issues
if (items.length > 3) {
    console.log('Array has more than 3 items');
} else if (items.length === 3) {
    console.log('Array has exactly 3 items');
} else {
    console.log('Array has fewer than 3 items');
}

// Loop with potential issues
for (let i = 0; i < items.length; i++) {
    console.log(`Item ${i}: ${items[i]}`);
}

// Promise with potential issues
const fetchData = () => {
    return new Promise((resolve, reject) => {
        setTimeout(() => {
            if (Math.random() > 0.5) {
                resolve({ data: 'Success', timestamp: Date.now() });
            } else {
                reject(new Error('Random failure'));
            }
        }, 1000);
    });
};

// Async/await with potential issues
const main = async () => {
    try {
        const result = await fetchData();
        console.log('Data received:', result);

        const processed = processData([{ id: 1, name: 'test' }]);
        console.log('Processed data:', processed);

        greet(name);

        // Unhandled promise rejection
        fetchData().then(data => {
            console.log('Unhandled:', data);
        });

    } catch (error) {
        console.error('Error:', error.message);
        // Missing rethrow or proper error handling
    }

    // Missing return statement
};

// Export with potential issues
module.exports = {
    greet,
    processData,
    fetchData,
    config,
    unusedExport: function() { return 'never used'; },
};

// Call main function
main();

// Global variable pollution
GLOBAL_VAR = 'This pollutes global scope';
window.anotherGlobal = 'Another global variable';

// Console statements in production code
console.log('Debug info');
console.warn('Warning message');
console.error('Error message');

// Eval usage (security issue)
const evilEval = eval('console.log("Eval is evil")');

// setTimeout with string (security issue)
setTimeout('console.log("Bad setTimeout")', 1000);

File: .config/nvim/testing/sample.lua
-- Sample Lua file with various issues for testing

-- Variables with potential issues
local name = "World"
local count = 5
local unused_var = "This variable is not used"
local another_unused = 42

-- Global variable (should be local)
global_var = "This should be local"

-- Function with potential issues
local function greet(name)
    print("Hello, " .. name .. "!")
    print("Count is: " .. count)

    -- Use of undeclared variable
    print(undeclared_var)

    -- Missing local keyword
    message = "This should be local"

    -- Unused parameter
    local function with_unused_param(param1, param2, param3)
        return param1
    end

    return message
end

-- Table with potential issues
local config = {
    apiUrl = "https://api.example.com",
    timeout = 5000,
    retries = 3,
    -- Trailing comma
}

-- Table with duplicate keys
local bad_config = {
    url = "https://example.com",
    timeout = 3000,
    url = "https://backup.com", -- Duplicate key
}

-- Array with potential issues
local items = {1, 2, 3, 4, 5}
local mixedArray = {"string", 42, true, nil, {key = "value"}}
local empty_table = {}

-- Conditional with potential issues
if #items > 3 then
    print("Array has more than 3 items")
elseif #items == 3 then
    print("Array has exactly 3 items")
else
    print("Array has fewer than 3 items")
end

-- Loop with potential issues
for i = 1, #items do
    print("Item " .. i .. ": " .. items[i])
end

-- Pairs loop with potential issues
for key, value in pairs(config) do
    print("Config " .. key .. ": " .. tostring(value))
end

-- Class-like structure with potential issues
local Person = {}
Person.__index = Person

function Person.new(name, age)
    local self = setmetatable({}, Person)
    self.name = name
    self.age = age
    return self
end

function Person:greet(greeting)
    greeting = greeting or "Hello"
    return greeting .. ", " .. self.name .. "!"
end

function Person:to_dict()
    return {
        name = self.name,
        age = self.age,
        greeting = self:greet()
    }
end

-- Module pattern with potential issues
local Utils = {}

function Utils.split(str, delimiter)
    local result = {}
    local pattern = string.format("([^%s]+)", delimiter)

    for match in string.gmatch(str, pattern) do
        table.insert(result, match)
    end

    return result
end

function Utils.join(tbl, delimiter)
    local result = ""
    for i, item in ipairs(tbl) do
        if i > 1 then
            result = result .. delimiter
        end
        result = result .. tostring(item)
    end
    return result
end

-- Error handling with potential issues
local function safe_divide(a, b)
    if b == 0 then
        error("Division by zero!")
    end
    return a / b
end

-- File operations with potential issues
local function write_sample_file()
    local file = io.open("sample.txt", "w")
    if not file then
        print("Failed to open file for writing")
        return
    end

    file:write("Sample Lua output\n")
    file:write("Generated at: " .. os.date() .. "\n")

    -- Missing file close
    -- file:close()
end

-- Function that's never called
local never_called = function()
    print("This function is never called")
end

-- Global function pollution
function pollute_global_scope()
    return "This pollutes global scope"
end

-- String manipulation with potential issues
local function process_text(text)
    -- Multiple string operations
    local upper = string.upper(text)
    local lower = string.lower(text)
    local reversed = string.reverse(text)

    return {
        original = text,
        upper = upper,
        lower = lower,
        reversed = reversed,
        length = string.len(text)
    }
end

-- Main execution with potential issues
local function main()
    -- Create person
    local person = Person.new("Alice", 25)
    print(person:greet("Hi there"))

    -- Process text
    local text = "Hello, World!"
    local processed = process_text(text)

    for key, value in pairs(processed) do
        print(key .. ": " .. tostring(value))
    end

    -- Test utils
    local parts = Utils.split("a,b,c,d", ",")
    local joined = Utils.join(parts, "-")
    print("Split and joined: " .. joined)

    -- Safe division
    local success, result = pcall(safe_divide, 10, 2)
    if success then
        print("Division result: " .. result)
    else
        print("Division error: " .. result)
    end

    -- Write file
    write_sample_file()

    -- Use of global variable
    global_var = "This should be local"
    print(global_var)

    -- Unused variable in main
    local main_unused = "This is not used"

    -- Missing return statement
end

-- Call main function
main()

-- Export for module usage
return {
    Person = Person,
    Utils = Utils,
    greet = greet,
    process_text = process_text
}
